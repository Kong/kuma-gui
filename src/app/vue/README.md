---
section: Packages
---
# vue

Automatic `vue` initialization using our projects dependency injection layer.

