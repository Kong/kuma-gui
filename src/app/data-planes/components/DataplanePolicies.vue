<template>
  <LoadingBlock v-if="isLoading" />

  <ErrorBlock
    v-else-if="error !== null"
    :error="error"
  />

  <div
    v-else-if="policyTypeEntries.length > 0"
    class="policies-list"
  >
    <SidecarDataplanePolicyList
      :dpp-name="props.dataplaneOverview.name"
      :policy-type-entries="policyTypeEntries"
      :rule-entries="ruleEntries"
    />
  </div>

  <div
    v-else-if="meshGatewayListenerEntries.length > 0 && meshGatewayDataplane !== null"
    class="policies-list"
  >
    <MeshGatewayDataplanePolicyList
      :mesh-gateway-dataplane="meshGatewayDataplane"
      :mesh-gateway-listener-entries="meshGatewayListenerEntries"
      :mesh-gateway-route-policies="meshGatewayRoutePolicies"
    />
  </div>

  <EmptyBlock v-else />
</template>

<script lang="ts" setup>
import { PropType, ref, watch } from 'vue'

import MeshGatewayDataplanePolicyList from './MeshGatewayDataplanePolicyList.vue'
import SidecarDataplanePolicyList from './SidecarDataplanePolicyList.vue'
import EmptyBlock from '@/app/common/EmptyBlock.vue'
import ErrorBlock from '@/app/common/ErrorBlock.vue'
import LoadingBlock from '@/app/common/LoadingBlock.vue'
import {
  DataPlaneOverview,
  DataplaneRule,
  LabelValue,
  MeshGatewayDataplane,
  MeshGatewayListenerEntry,
  MeshGatewayRouteEntry,
  MeshGatewayRoutePolicy,
  MatchedPolicyType,
  PolicyTypeEntry,
  PolicyTypeEntryConnection,
  PolicyTypeEntryOrigin,
  RuleEntry,
  RuleEntryConnection,
  SidecarDataplane,
} from '@/types/index.d'
import { useKumaApi } from '@/utilities'
import { toYaml } from '@/utilities/toYaml'

const kumaApi = useKumaApi()

const props = defineProps({
  dataplaneOverview: {
    type: Object as PropType<DataPlaneOverview>,
    required: true,
  },
})

const meshGatewayDataplane = ref<MeshGatewayDataplane | null>(null)
const policyTypeEntries = ref<PolicyTypeEntry[]>([])
const ruleEntries = ref<RuleEntry[]>([])
const meshGatewayListenerEntries = ref<MeshGatewayListenerEntry[]>([])
const meshGatewayRoutePolicies = ref<MeshGatewayRoutePolicy[]>([])
const isLoading = ref(true)
const error = ref<Error | null>(null)

watch(() => props.dataplaneOverview.name, function () {
  fetchPolicies()
})

fetchPolicies()

async function fetchPolicies(): Promise<void> {
  error.value = null
  isLoading.value = true
  policyTypeEntries.value = []
  ruleEntries.value = []
  meshGatewayListenerEntries.value = []
  meshGatewayRoutePolicies.value = []

  try {
    const isMeshGatewayDataplane = props.dataplaneOverview.dataplane.networking.gateway?.type?.toUpperCase() === 'BUILTIN'

    if (isMeshGatewayDataplane) {
      meshGatewayDataplane.value = await kumaApi.getMeshGatewayDataplane({
        mesh: props.dataplaneOverview.mesh,
        name: props.dataplaneOverview.name,
      })

      meshGatewayListenerEntries.value = getMeshGatewayListenerEntries(meshGatewayDataplane.value)
      meshGatewayRoutePolicies.value = getPolicyRoutes(meshGatewayDataplane.value.policies)
    } else {
      const { items: sidecarDataplanes } = await kumaApi.getSidecarDataplanePolicies({
        mesh: props.dataplaneOverview.mesh,
        name: props.dataplaneOverview.name,
      })

      policyTypeEntries.value = getPolicyTypeEntries(sidecarDataplanes ?? [])

      const { items: rules } = await kumaApi.getDataplaneRules({
        mesh: props.dataplaneOverview.mesh,
        name: props.dataplaneOverview.name,
      })

      ruleEntries.value = getRuleEntries(rules ?? [])
    }
  } catch (err) {
    if (err instanceof Error) {
      error.value = err
    } else {
      console.error(err)
    }
  } finally {
    isLoading.value = false
  }
}

function getMeshGatewayListenerEntries(meshGatewayDataplane: MeshGatewayDataplane): MeshGatewayListenerEntry[] {
  const meshGatewayListenerEntries: MeshGatewayListenerEntry[] = []

  const listeners = meshGatewayDataplane.listeners ?? []
  for (const listener of listeners) {
    for (const host of listener.hosts) {
      for (const route of host.routes) {
        const routeEntries: MeshGatewayRouteEntry[] = []

        for (const destination of route.destinations) {
          const policies = getPolicyRoutes(destination.policies)

          const routeEntry: MeshGatewayRouteEntry = {
            routeName: route.route,
            route: {
              name: 'policy-detail-view',
              params: {
                mesh: meshGatewayDataplane.gateway.mesh,
                policyType: 'MeshGatewayRoute',
                policy: route.route,
              },
            },
            service: destination.tags['kuma.io/service'],
            policies,
          }

          routeEntries.push(routeEntry)
        }

        meshGatewayListenerEntries.push({
          protocol: listener.protocol,
          port: listener.port,
          hostName: host.hostName,
          routeEntries,
        })
      }
    }
  }

  return meshGatewayListenerEntries
}

function getPolicyRoutes(policies: Record<string, MatchedPolicyType> | undefined): MeshGatewayRoutePolicy[] {
  if (policies === undefined) {
    return []
  }

  const policyRoutes: MeshGatewayRoutePolicy[] = []

  for (const policy of Object.values(policies)) {
    policyRoutes.push({
      type: policy.type,
      name: policy.name,
      route: {
        name: 'policy-detail-view',
        params: {
          mesh: policy.mesh,
          policyType: policy.type,
          policy: policy.name,
        },
      },
    })
  }

  return policyRoutes
}

/**
 * Transforms `SidecarDataplane` objects into policy type entries which are going to be displayed in this view.
 */
function getPolicyTypeEntries(sidecarDataplanes: SidecarDataplane[]): PolicyTypeEntry[] {
  // Uses a `Map` to store entries by type so they can be retrieved and updated while iterating over the `SidecarDataplane` objects.
  const policyTypeEntriesByType = new Map<string, PolicyTypeEntry>()

  for (const sidecarDataplane of sidecarDataplanes) {
    const { type, service } = sidecarDataplane

    // The `service` field, when set, represents the name of the destination service of traffic.
    const destinationTags: LabelValue[] = typeof service === 'string' && service !== '' ? [{ label: 'kuma.io/service', value: service }] : []
    const name = type === 'inbound' || type === 'outbound' ? sidecarDataplane.name : null

    for (const [policyTypeName, policies] of Object.entries(sidecarDataplane.matchedPolicies)) {
      if (!policyTypeEntriesByType.has(policyTypeName)) {
        policyTypeEntriesByType.set(policyTypeName, {
          type: policyTypeName,
          connections: [],
        })
      }

      const policyTypeEntry = policyTypeEntriesByType.get(policyTypeName) as PolicyTypeEntry

      for (const policy of policies) {
        const connections = getPolicyTypeEntryConnections(policy, sidecarDataplane, destinationTags, name)

        policyTypeEntry.connections.push(...connections)
      }
    }
  }

  const policyTypeEntries = Array.from(policyTypeEntriesByType.values())

  policyTypeEntries.sort((policyTypeEntryA, policyTypeEntryB) => policyTypeEntryA.type.localeCompare(policyTypeEntryB.type))

  return policyTypeEntries
}

function getPolicyTypeEntryConnections(policy: MatchedPolicyType, sidecarDataplane: SidecarDataplane, destinationTags: LabelValue[], name: string | null): PolicyTypeEntryConnection[] {
  const config = policy.conf && Object.keys(policy.conf).length > 0 ? toYaml(policy.conf) : null
  const origin: PolicyTypeEntryOrigin = {
    name: policy.name,
    route: {
      name: 'policy-detail-view',
      params: {
        mesh: policy.mesh,
        policyType: policy.type,
        policy: policy.name,
      },
    },
  }
  const origins: PolicyTypeEntryOrigin[] = [origin]

  const policyTypeEntryConnections: PolicyTypeEntryConnection[] = []

  if (sidecarDataplane.type === 'inbound' && Array.isArray(policy.sources)) {
    for (const { match } of policy.sources) {
      const sourceTags: LabelValue[] = [{ label: 'kuma.io/service', value: match['kuma.io/service'] }]
      const connection: PolicyTypeEntryConnection = { sourceTags, destinationTags, name, config, origins }

      policyTypeEntryConnections.push(connection)
    }
  } else {
    const sourceTags: LabelValue[] = []
    const connection: PolicyTypeEntryConnection = { sourceTags, destinationTags, name, config, origins }

    policyTypeEntryConnections.push(connection)
  }

  return policyTypeEntryConnections
}

/**
 * Transforms `DataplaneRule` objects into rule entries which are going to be displayed in this view.
 */
function getRuleEntries(rules: DataplaneRule[]): RuleEntry[] {
  // Uses a `Map` to store entries by type so they can be retrieved and updated while iterating over the rules.
  const policyTypeEntriesByType = new Map<string, RuleEntry>()

  for (const rule of rules) {
    if (!policyTypeEntriesByType.has(rule.policyType)) {
      policyTypeEntriesByType.set(rule.policyType, {
        type: rule.policyType,
        connections: [],
      })
    }

    const policyTypeEntry = policyTypeEntriesByType.get(rule.policyType) as RuleEntry
    const connections = getRuleEntryConnections(rule)

    policyTypeEntry.connections.push(...connections)
  }

  const policyTypeEntries = Array.from(policyTypeEntriesByType.values())

  policyTypeEntries.sort((policyTypeEntryA, policyTypeEntryB) => policyTypeEntryA.type.localeCompare(policyTypeEntryB.type))

  return policyTypeEntries
}

function getRuleEntryConnections(rule: DataplaneRule): RuleEntryConnection[] {
  const { type, service, subset, conf } = rule

  // Guards against likely API changes. The response currently contains `"subset": {}` instead of omitting the value or setting it to `null`, but that might change in the future.
  const subsetEntries = subset ? Object.entries(subset) : []
  let sourceTags: LabelValue[]
  let destinationTags: LabelValue[]

  // Determines source tags.
  if (type === 'ClientSubset') {
    if (subsetEntries.length > 0) {
    // For client subsets, the source is represented by `subset` (i.e. tags)
      sourceTags = subsetEntries.map(([label, value]) => ({ label, value }))
    } else {
      // Sets the wildcard service tag for client subsets without subset values.
      sourceTags = [
        {
          label: 'kuma.io/service',
          value: '*',
        },
      ]
    }
  } else {
    // For destination subsets and single items, the source is the DPP for which we don’t want to show anything.
    sourceTags = []
  }

  // Determines destination tags.
  if (type === 'DestinationSubset') {
    // For destination subsets, the destination is represented by either `subset` (which has priority) or `service`. `subset` is more specific than `service` so it needs to be handled first.
    if (subsetEntries.length > 0) {
      destinationTags = subsetEntries.map(([label, value]) => ({ label, value }))
    } else if (typeof service === 'string' && service !== '') {
      // The `service` field, when set, represents the name of the destination service of traffic.
      destinationTags = [
        {
          label: 'kuma.io/service',
          value: service,
        },
      ]
    } else {
      // For destination subsets with empty or absent `subset` field and empty or absent `service`, we can set the wildcard service tag to indicate that the traffic goes to all services.
      destinationTags = [
        {
          label: 'kuma.io/service',
          value: '*',
        },
      ]
    }
  } else if (type === 'ClientSubset' && typeof service === 'string' && service !== '') {
    // The `service` field, when set, represents the name of the destination service of traffic.
    destinationTags = [
      {
        label: 'kuma.io/service',
        value: service,
      },
    ]
  } else {
    destinationTags = []
  }

  const addresses = rule.addresses ?? []
  const config = conf && Object.keys(conf).length > 0 ? toYaml(conf) : null
  const origins: PolicyTypeEntryOrigin[] = []

  for (const ruleOrigin of rule.origins) {
    origins.push({
      name: ruleOrigin.name,
      route: {
        name: 'policy-detail-view',
        params: {
          mesh: ruleOrigin.mesh,
          policyType: rule.policyType,
          policy: ruleOrigin.name,
        },
      },
    })
  }

  const connection: RuleEntryConnection = { type: { sourceTags, destinationTags }, addresses, config, origins }

  return [connection]
}
</script>

<style lang="scss" scoped>
.policies-list {
  padding: var(--spacing-sm);
}
</style>
