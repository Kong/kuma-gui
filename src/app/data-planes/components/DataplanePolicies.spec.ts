import { RouterLinkStub } from '@vue/test-utils'
import { rest } from 'msw'
import { render, screen } from '@testing-library/vue'
import userEvent from '@testing-library/user-event'

import DataplanePolicies from './DataplanePolicies.vue'
import { store, storeKey } from '@/store/store'
import { server } from '@/jest-setup'

async function renderComponent(props = {}) {
  await store.dispatch('fetchPolicies')

  return render(DataplanePolicies, {
    global: {
      plugins: [[store, storeKey]],
      stubs: {
        'router-link': RouterLinkStub,
      },
    },
    props,
  })
}

describe('DataplanePolicies.vue', () => {
  it('renders snapshot', async () => {
    const { container } = await renderComponent({
      dataPlane: {
        mesh: 'foo',
        name: 'dataplane-test-456',
        networking: {},
      },
    })

    await userEvent.click(await screen.findByText('web'))

    expect(container).toMatchSnapshot()
  })

  it('renders loading', async () => {
    server.use(
      rest.get(import.meta.env.VITE_KUMA_API_SERVER_URL + 'meshes/:mesh/dataplanes/:dataplaneName/policies', (req, res, ctx) =>
        res(ctx.status(200), ctx.json({ total: 0, items: [] })),
      ),
    )

    await renderComponent({
      dataPlane: {
        mesh: 'foo',
        name: 'dataplane-test-456',
        networking: {},
      },
    })

    expect(screen.getByTestId('loading-block')).toBeInTheDocument()
  })

  it('renders error', async () => {
    jest.spyOn(console, 'error').mockImplementation(() => { }) // silence console errors

    server.use(
      rest.get(import.meta.env.VITE_KUMA_API_SERVER_URL + 'meshes/:mesh/dataplanes/:dataplaneName/policies', (req, res, ctx) =>
        res(ctx.status(500), ctx.json({})),
      ),
    )

    await renderComponent({
      dataPlane: {
        mesh: 'foo',
        name: 'dataplane-test-456',
        networking: {},
      },
    })

    expect((await screen.findAllByText(/An error has occurred while trying to load this data./))[0]).toBeInTheDocument()
  })

  it('renders no item', async () => {
    server.use(
      rest.get(import.meta.env.VITE_KUMA_API_SERVER_URL + 'meshes/:mesh/dataplanes/:dataplaneName/policies', (req, res, ctx) =>
        res(ctx.status(200), ctx.json({ total: 0, items: [] })),
      ),
    )

    await renderComponent({
      dataPlane: {
        mesh: 'foo',
        name: 'dataplane-test-456',
        networking: {},
      },
    })

    expect(await screen.findByText(/There is no data to display./)).toBeInTheDocument()
  })
})
