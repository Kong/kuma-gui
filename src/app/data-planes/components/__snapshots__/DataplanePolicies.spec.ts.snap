// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataplanePolicies.vue renders snapshot 1`] = `
<div
  class="policies-list"
>
  
  <h2
    class="visually-hidden"
  >
     Policies 
  </h2>
  <ul
    class="accordion-list"
  >
    
    
    <li
      class="accordion-item active"
    >
      <button
        aria-expanded="true"
        class="accordion-item-header"
        data-testid="accordion-item-button"
        type="button"
      >
        
        <h3
          class="policy-type-heading"
        >
          <span
            class="policy-type-tag"
          >
            <img
              alt=""
              class="policy-type-tag-icon"
              src=""
            />
            
            FaultInjection (3) 
            
          </span>
        </h3>
        
      </button>
      <transition-stub
        appear="false"
        css="true"
        name="accordion"
        persisted="false"
      >
        <div
          class="accordion-item-content"
          data-testid="accordion-item-content"
        >
          
          <div
            class="policy-list"
          >
            <div
              class="k-table-container policy-type-table"
              is-clickable=""
            >
              <!---->
              <section
                class="k-table-wrapper"
              >
                <table
                  class="k-table has-hover"
                  data-tableid="aaaabbbb-cccc-dddd-eeee-ffffffffffff"
                >
                  <thead
                    class=""
                  >
                    <tr
                      class=""
                    >
                      
                      <th
                        class=""
                      >
                        <span
                          class="d-flex align-items-center"
                        >
                          
                          <span
                            class=""
                          >
                            From
                          </span>
                          
                          <!---->
                        </span>
                      </th>
                      <th
                        class=""
                      >
                        <span
                          class="d-flex align-items-center"
                        >
                          
                          <span
                            class=""
                          >
                            To
                          </span>
                          
                          <!---->
                        </span>
                      </th>
                      <th
                        class=""
                      >
                        <span
                          class="d-flex align-items-center"
                        >
                          
                          <span
                            class=""
                          >
                            On
                          </span>
                          
                          <!---->
                        </span>
                      </th>
                      <th
                        class=""
                      >
                        <span
                          class="d-flex align-items-center"
                        >
                          
                          <span
                            class=""
                          >
                            Conf
                          </span>
                          
                          <!---->
                        </span>
                      </th>
                      <th
                        class=""
                      >
                        <span
                          class="d-flex align-items-center"
                        >
                          
                          <span
                            class=""
                          >
                            Origin policies
                          </span>
                          
                          <!---->
                        </span>
                      </th>
                      
                    </tr>
                  </thead>
                  <tbody>
                    
                    <tr>
                      
                      <td
                        class="cell-sourceTags"
                      >
                        
                        <span
                          class="tag-list"
                        >
                          
                          <div
                            class="k-badge d-inline-flex k-badge-default k-badge-rounded tag-badge"
                            tabindex="0"
                          >
                            <div
                              class="k-badge-text truncate"
                            >
                              <div
                                class="k-badge-text truncate"
                              >
                                
                                <a
                                  class=""
                                  href="/mesh/default/service/service-a"
                                >
                                  kuma.io/service:
                                  <b>
                                    service-a
                                  </b>
                                </a>
                                
                              </div>
                            </div>
                            <!---->
                          </div>
                          
                        </span>
                        
                      </td>
                      <td
                        class="cell-destinationTags"
                      >
                        
                        <span
                          class="tag-list"
                        >
                          
                          <div
                            class="k-badge d-inline-flex k-badge-default k-badge-rounded tag-badge"
                            tabindex="0"
                          >
                            <div
                              class="k-badge-text truncate"
                            >
                              <div
                                class="k-badge-text truncate"
                              >
                                
                                <a
                                  class=""
                                  href="/mesh/default/service/web"
                                >
                                  kuma.io/service:
                                  <b>
                                    web
                                  </b>
                                </a>
                                
                              </div>
                            </div>
                            <!---->
                          </div>
                          
                        </span>
                        
                      </td>
                      <td
                        class="cell-name"
                      >
                        
                        
                        192.168.0.1:80:81
                        
                        
                      </td>
                      <td
                        class="cell-config"
                      >
                        
                        <div
                          class="k-code-block theme-light code-block"
                          data-testid="k-code-block"
                          id="policies-0-0-code-block"
                          style="--maxLineNumberWidth: 1ch;"
                          tabindex="0"
                        >
                          <!---->
                          <div
                            class="k-code-block-content"
                          >
                            <pre
                              class="k-highlighted-code-block language-yaml"
                              data-testid="k-code-block-highlighted-code-block"
                              tabindex="0"
                            >
                                      
                              <span
                                class="k-line-number-rows"
                              >
                                
          
                                
                                <span
                                  class="k-line"
                                >
                                  
            
                                  <a
                                    class="k-line-anchor"
                                    id="policies-0-0-code-block-L1"
                                  >
                                    1
                                  </a>
                                  
          
                                </span>
                                <span
                                  class="k-line"
                                >
                                  
            
                                  <a
                                    class="k-line-anchor"
                                    id="policies-0-0-code-block-L2"
                                  >
                                    2
                                  </a>
                                  
          
                                </span>
                                <span
                                  class="k-line"
                                >
                                  
            
                                  <a
                                    class="k-line-anchor"
                                    id="policies-0-0-code-block-L3"
                                  >
                                    3
                                  </a>
                                  
          
                                </span>
                                
                                
        
                              </span>
                              
        
                              <code
                                class="language-yaml"
                              >
                                <span
                                  class="token key atrule"
                                >
                                  delay
                                </span>
                                <span
                                  class="token punctuation"
                                >
                                  :
                                </span>
                                
  
                                <span
                                  class="token key atrule"
                                >
                                  percentage
                                </span>
                                <span
                                  class="token punctuation"
                                >
                                  :
                                </span>
                                 
                                <span
                                  class="token number"
                                >
                                  90
                                </span>
                                
  
                                <span
                                  class="token key atrule"
                                >
                                  value
                                </span>
                                <span
                                  class="token punctuation"
                                >
                                  :
                                </span>
                                 5s
                              </code>
                              
      
                            </pre>
                            <div
                              class="k-code-block-secondary-actions"
                            >
                              <!---->
                              
                              
                            </div>
                          </div>
                        </div>
                        
                      </td>
                      <td
                        class="cell-origins"
                      >
                        
                        <ul
                          class="origin-list"
                        >
                          
                          <li>
                            <a
                              class=""
                              href="/mesh/default/policy/fault-injections/fi1.kuma-system"
                            >
                              fi1.kuma-system
                            </a>
                          </li>
                          
                        </ul>
                        
                      </td>
                      
                    </tr>
                    <tr>
                      
                      <td
                        class="cell-sourceTags"
                      >
                        
                        <span
                          class="tag-list"
                        >
                          
                          <div
                            class="k-badge d-inline-flex k-badge-default k-badge-rounded tag-badge"
                            tabindex="0"
                          >
                            <div
                              class="k-badge-text truncate"
                            >
                              <div
                                class="k-badge-text truncate"
                              >
                                
                                <a
                                  class=""
                                  href="/mesh/default/service/service-b"
                                >
                                  kuma.io/service:
                                  <b>
                                    service-b
                                  </b>
                                </a>
                                
                              </div>
                            </div>
                            <!---->
                          </div>
                          
                        </span>
                        
                      </td>
                      <td
                        class="cell-destinationTags"
                      >
                        
                        <span
                          class="tag-list"
                        >
                          
                          <div
                            class="k-badge d-inline-flex k-badge-default k-badge-rounded tag-badge"
                            tabindex="0"
                          >
                            <div
                              class="k-badge-text truncate"
                            >
                              <div
                                class="k-badge-text truncate"
                              >
                                
                                <a
                                  class=""
                                  href="/mesh/default/service/web"
                                >
                                  kuma.io/service:
                                  <b>
                                    web
                                  </b>
                                </a>
                                
                              </div>
                            </div>
                            <!---->
                          </div>
                          
                        </span>
                        
                      </td>
                      <td
                        class="cell-name"
                      >
                        
                        
                        192.168.0.1:80:81
                        
                        
                      </td>
                      <td
                        class="cell-config"
                      >
                        
                        <div
                          class="k-code-block theme-light code-block"
                          data-testid="k-code-block"
                          id="policies-0-1-code-block"
                          style="--maxLineNumberWidth: 1ch;"
                          tabindex="0"
                        >
                          <!---->
                          <div
                            class="k-code-block-content"
                          >
                            <pre
                              class="k-highlighted-code-block language-yaml"
                              data-testid="k-code-block-highlighted-code-block"
                              tabindex="0"
                            >
                                      
                              <span
                                class="k-line-number-rows"
                              >
                                
          
                                
                                <span
                                  class="k-line"
                                >
                                  
            
                                  <a
                                    class="k-line-anchor"
                                    id="policies-0-1-code-block-L1"
                                  >
                                    1
                                  </a>
                                  
          
                                </span>
                                <span
                                  class="k-line"
                                >
                                  
            
                                  <a
                                    class="k-line-anchor"
                                    id="policies-0-1-code-block-L2"
                                  >
                                    2
                                  </a>
                                  
          
                                </span>
                                <span
                                  class="k-line"
                                >
                                  
            
                                  <a
                                    class="k-line-anchor"
                                    id="policies-0-1-code-block-L3"
                                  >
                                    3
                                  </a>
                                  
          
                                </span>
                                
                                
        
                              </span>
                              
        
                              <code
                                class="language-yaml"
                              >
                                <span
                                  class="token key atrule"
                                >
                                  delay
                                </span>
                                <span
                                  class="token punctuation"
                                >
                                  :
                                </span>
                                
  
                                <span
                                  class="token key atrule"
                                >
                                  percentage
                                </span>
                                <span
                                  class="token punctuation"
                                >
                                  :
                                </span>
                                 
                                <span
                                  class="token number"
                                >
                                  90
                                </span>
                                
  
                                <span
                                  class="token key atrule"
                                >
                                  value
                                </span>
                                <span
                                  class="token punctuation"
                                >
                                  :
                                </span>
                                 5s
                              </code>
                              
      
                            </pre>
                            <div
                              class="k-code-block-secondary-actions"
                            >
                              <!---->
                              
                              
                            </div>
                          </div>
                        </div>
                        
                      </td>
                      <td
                        class="cell-origins"
                      >
                        
                        <ul
                          class="origin-list"
                        >
                          
                          <li>
                            <a
                              class=""
                              href="/mesh/default/policy/fault-injections/fi1.kuma-system"
                            >
                              fi1.kuma-system
                            </a>
                          </li>
                          
                        </ul>
                        
                      </td>
                      
                    </tr>
                    <tr>
                      
                      <td
                        class="cell-sourceTags"
                      >
                        
                        <span
                          class="tag-list"
                        >
                          
                          <div
                            class="k-badge d-inline-flex k-badge-default k-badge-rounded tag-badge"
                            tabindex="0"
                          >
                            <div
                              class="k-badge-text truncate"
                            >
                              <div
                                class="k-badge-text truncate"
                              >
                                
                                <span>
                                  kuma.io/service:
                                  <b>
                                    *
                                  </b>
                                </span>
                                
                              </div>
                            </div>
                            <!---->
                          </div>
                          
                        </span>
                        
                      </td>
                      <td
                        class="cell-destinationTags"
                      >
                        
                        <span
                          class="tag-list"
                        >
                          
                          <div
                            class="k-badge d-inline-flex k-badge-default k-badge-rounded tag-badge"
                            tabindex="0"
                          >
                            <div
                              class="k-badge-text truncate"
                            >
                              <div
                                class="k-badge-text truncate"
                              >
                                
                                <a
                                  class=""
                                  href="/mesh/default/service/web"
                                >
                                  kuma.io/service:
                                  <b>
                                    web
                                  </b>
                                </a>
                                
                              </div>
                            </div>
                            <!---->
                          </div>
                          
                        </span>
                        
                      </td>
                      <td
                        class="cell-name"
                      >
                        
                        
                        192.168.0.1:80:81
                        
                        
                      </td>
                      <td
                        class="cell-config"
                      >
                        
                        <div
                          class="k-code-block theme-light code-block"
                          data-testid="k-code-block"
                          id="policies-0-2-code-block"
                          style="--maxLineNumberWidth: 1ch;"
                          tabindex="0"
                        >
                          <!---->
                          <div
                            class="k-code-block-content"
                          >
                            <pre
                              class="k-highlighted-code-block language-yaml"
                              data-testid="k-code-block-highlighted-code-block"
                              tabindex="0"
                            >
                                      
                              <span
                                class="k-line-number-rows"
                              >
                                
          
                                
                                <span
                                  class="k-line"
                                >
                                  
            
                                  <a
                                    class="k-line-anchor"
                                    id="policies-0-2-code-block-L1"
                                  >
                                    1
                                  </a>
                                  
          
                                </span>
                                <span
                                  class="k-line"
                                >
                                  
            
                                  <a
                                    class="k-line-anchor"
                                    id="policies-0-2-code-block-L2"
                                  >
                                    2
                                  </a>
                                  
          
                                </span>
                                <span
                                  class="k-line"
                                >
                                  
            
                                  <a
                                    class="k-line-anchor"
                                    id="policies-0-2-code-block-L3"
                                  >
                                    3
                                  </a>
                                  
          
                                </span>
                                
                                
        
                              </span>
                              
        
                              <code
                                class="language-yaml"
                              >
                                <span
                                  class="token key atrule"
                                >
                                  abort
                                </span>
                                <span
                                  class="token punctuation"
                                >
                                  :
                                </span>
                                
  
                                <span
                                  class="token key atrule"
                                >
                                  percentage
                                </span>
                                <span
                                  class="token punctuation"
                                >
                                  :
                                </span>
                                 
                                <span
                                  class="token number"
                                >
                                  80
                                </span>
                                
  
                                <span
                                  class="token key atrule"
                                >
                                  httpStatus
                                </span>
                                <span
                                  class="token punctuation"
                                >
                                  :
                                </span>
                                 
                                <span
                                  class="token number"
                                >
                                  500
                                </span>
                              </code>
                              
      
                            </pre>
                            <div
                              class="k-code-block-secondary-actions"
                            >
                              <!---->
                              
                              
                            </div>
                          </div>
                        </div>
                        
                      </td>
                      <td
                        class="cell-origins"
                      >
                        
                        <ul
                          class="origin-list"
                        >
                          
                          <li>
                            <a
                              class=""
                              href="/mesh/default/policy/fault-injections/web-to-backend.kuma-system"
                            >
                              web-to-backend.kuma-system
                            </a>
                          </li>
                          
                        </ul>
                        
                      </td>
                      
                    </tr>
                    
                  </tbody>
                </table>
                <!---->
              </section>
            </div>
          </div>
          
        </div>
      </transition-stub>
    </li>
    <li
      class="accordion-item"
    >
      <button
        aria-expanded="false"
        class="accordion-item-header"
        data-testid="accordion-item-button"
        type="button"
      >
        
        <h3
          class="policy-type-heading"
        >
          <span
            class="policy-type-tag"
          >
            <img
              alt=""
              class="policy-type-tag-icon"
              src=""
            />
            
            HealthCheck (4) 
            
          </span>
        </h3>
        
      </button>
      <transition-stub
        appear="false"
        css="true"
        name="accordion"
        persisted="false"
      >
        <!--v-if-->
      </transition-stub>
    </li>
    <li
      class="accordion-item"
    >
      <button
        aria-expanded="false"
        class="accordion-item-header"
        data-testid="accordion-item-button"
        type="button"
      >
        
        <h3
          class="policy-type-heading"
        >
          <span
            class="policy-type-tag"
          >
            <img
              alt=""
              class="policy-type-tag-icon"
              src=""
            />
            
            Timeout (1) 
            
          </span>
        </h3>
        
      </button>
      <transition-stub
        appear="false"
        css="true"
        name="accordion"
        persisted="false"
      >
        <!--v-if-->
      </transition-stub>
    </li>
    <li
      class="accordion-item"
    >
      <button
        aria-expanded="false"
        class="accordion-item-header"
        data-testid="accordion-item-button"
        type="button"
      >
        
        <h3
          class="policy-type-heading"
        >
          <span
            class="policy-type-tag"
          >
            <img
              alt=""
              class="policy-type-tag-icon"
              src=""
            />
            
            TrafficPermission (1) 
            
          </span>
        </h3>
        
      </button>
      <transition-stub
        appear="false"
        css="true"
        name="accordion"
        persisted="false"
      >
        <!--v-if-->
      </transition-stub>
    </li>
    <li
      class="accordion-item"
    >
      <button
        aria-expanded="false"
        class="accordion-item-header"
        data-testid="accordion-item-button"
        type="button"
      >
        
        <h3
          class="policy-type-heading"
        >
          <span
            class="policy-type-tag"
          >
            <img
              alt=""
              class="policy-type-tag-icon"
              src=""
            />
            
            TrafficTrace (1) 
            
          </span>
        </h3>
        
      </button>
      <transition-stub
        appear="false"
        css="true"
        name="accordion"
        persisted="false"
      >
        <!--v-if-->
      </transition-stub>
    </li>
    
    
  </ul>
  <div
    class="mt-2"
  >
    <h2>
      Rules
    </h2>
    <ul
      class="accordion-list"
    >
      
      
      <li
        class="accordion-item"
      >
        <button
          aria-expanded="false"
          class="accordion-item-header"
          data-testid="accordion-item-button"
          type="button"
        >
          
          <h3
            class="policy-type-heading"
          >
            <span
              class="policy-type-tag"
            >
              <img
                alt=""
                class="policy-type-tag-icon"
                src=""
              />
              
              MeshCircuitBreaker (6) 
              
            </span>
          </h3>
          
        </button>
        <transition-stub
          appear="false"
          css="true"
          name="accordion"
          persisted="false"
        >
          <!--v-if-->
        </transition-stub>
      </li>
      <li
        class="accordion-item"
      >
        <button
          aria-expanded="false"
          class="accordion-item-header"
          data-testid="accordion-item-button"
          type="button"
        >
          
          <h3
            class="policy-type-heading"
          >
            <span
              class="policy-type-tag"
            >
              <img
                alt=""
                class="policy-type-tag-icon"
                src=""
              />
              
              MeshRateLimit (1) 
              
            </span>
          </h3>
          
        </button>
        <transition-stub
          appear="false"
          css="true"
          name="accordion"
          persisted="false"
        >
          <!--v-if-->
        </transition-stub>
      </li>
      <li
        class="accordion-item"
      >
        <button
          aria-expanded="false"
          class="accordion-item-header"
          data-testid="accordion-item-button"
          type="button"
        >
          
          <h3
            class="policy-type-heading"
          >
            <span
              class="policy-type-tag"
            >
              <img
                alt=""
                class="policy-type-tag-icon"
                src=""
              />
              
              MeshRetry (2) 
              
            </span>
          </h3>
          
        </button>
        <transition-stub
          appear="false"
          css="true"
          name="accordion"
          persisted="false"
        >
          <!--v-if-->
        </transition-stub>
      </li>
      <li
        class="accordion-item"
      >
        <button
          aria-expanded="false"
          class="accordion-item-header"
          data-testid="accordion-item-button"
          type="button"
        >
          
          <h3
            class="policy-type-heading"
          >
            <span
              class="policy-type-tag"
            >
              <img
                alt=""
                class="policy-type-tag-icon"
                src=""
              />
              
              MeshTrace (1) 
              
            </span>
          </h3>
          
        </button>
        <transition-stub
          appear="false"
          css="true"
          name="accordion"
          persisted="false"
        >
          <!--v-if-->
        </transition-stub>
      </li>
      
      
    </ul>
  </div>
  
</div>
`;
